Specify noDerivative and zeroDerivative. (#2864)
modelica/ModelicaSpecification: chapters/functions.tex

Assume that function \lstinline!f! takes a matrix and a scalar.
Since the matrix argument is usually a parameter expression it is then
useful to define the function as follows (the additional derivative =
\lstinline!fGeneralDer! is optional and can be used when the derivative of
the matrix or offset is non-zero). Note that \lstinline!fDer! must have

> \lstinline!zeroDerivative! for both \lstinline!y! and \lstinline!offset!, but \lstinline!fGeneralDer! shall not have

Reviewer: What does it mean: "must have zeroDerivative for both y and offset"? In the annotation for the derivative of fDer? Sorry, I realize that it is not really part of the PR because the text was there before. Just reviewing PRs brings attention to some things that I would otherwise disregard when simply reading. Should I open a separate issue?

> \lstinline!zeroDerivative! for either of them (it may \lstinline!zeroDerivative! for \lstinline!x_der!,

Reviewer: It looks like there is a verb missing in "(it may \lstinline!zeroDerivative! for \lstinline!x_der!" Same problem as the comment above: the sentence was there before this PR. Should I open a separate issue?

  //annotation(derivative(order=2) = f_general_der2);
end f_general_der;
\end{lstlisting}

> In the example above \lstinline!zeroDerivative=y! imply that

Author: Of course

> In the example above \lstinline!zeroDerivative=y! implies that

Reviewer: `offset` is missing:
```suggestion
In the example above \lstinline!zeroDerivative=y! and \lstinline!zeroDerivative=offset! imply that
```

\begin{eqnarray*}

> \frac{d}{dt}f(x(t),y(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}\\

Author: > I don't know if it's an issue with the example, with the formula, or with the definition of `noDerivative`. So for now I am just adding a comment here. May need to open a separate issue. The issue is with the `noDerivative` annotation. Looking at the example and the formula in the PR, `f_der` depends on `g`. The annotation "annotation(derivative(noDerivative = y) = f_der);" is specified on `f` without any information about `g`. How can one know which `g` `f_der` is valid for by just looking at `f` and the annotation?

You are right: It cannot.
We did at some point attempt to specify it by some more elaborate variant of noDerivative - but it was too complicated.

> \frac{d}{dt}f(x(t),y(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}\\

Author: > Should we include `offset` into the formula as well for consistency? Or remove it from the Modelica code for `f` for consistency and brevity?

I added it (but renamed to "o"). The reason to keep it is that it shows zeroDerivative for multiple variables.

> \frac{d}{dt}f(x(t),y(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}\\

Reviewer: > > I don't know if it's an issue with the example, with the formula, or with the definition of `noDerivative`. So for now I am just adding a comment here. May need to open a separate issue. The issue is with the `noDerivative` annotation. Looking at the example and the formula in the PR, `f_der` depends on `g`. The annotation "annotation(derivative(noDerivative = y) = f_der);" is specified on `f` without any information about `g`. How can one know which `g` `f_der` is valid for by just looking at `f` and the annotation?
> 
> You are right: It cannot.
> We did at some point attempt to specify it by some more elaborate variant of noDerivative - but it was too complicated.

That makes it a rather useless annotation then, in my opinion. If a tool does not know in which cases it should be used, the safest way is not to use it.

> \frac{d}{dt}f(x(t),y(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}\\

Author: On one hand it's true that it's unsafe and best to not use - on the other hand it's quite a common case for the functions Modelica.Media due to their construction. At some point we should investigate if that could be improved.

> \frac{d}{dt}f(x(t),y(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}\\

Reviewer: > On one hand it's true that it's unsafe and best to not use - on the other hand it's quite a common case for the functions Modelica.Media due to their construction. At some point we should investigate if that could be improved.

OK, I just want to double check that the formula and the example are correct and as intended. In the original response to my ticket #2840 there was no mention of `fg`. So I understood it that we always assume that `y` is a function of `g`. From this PR I see that `y` can be independent of `x` and `h` depends on what `g` is. So I just want to confirm that the way it is in the PR is the way it is meant to be.

I see that in Modelica.Media `Water.IF97_Utilities.T_props_ph`, for example, has a `derivative` annotation with `noDerivative` specified on the third argument, and that `T_props_ph` is always called  as `T_props_ph(p,h,waterBaseProp_ph(p,h,phase,region)`. So, I guess, the assumption is that whenever we need a derivative and it is called with the noDerivative argument as a function of other arguments we can use the derivative in the annotation?

> \frac{d}{dt}f(x(t),y(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}\\

Author: Yes.

> \frac{d}{dt}f(x(t),y(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}\\

Reviewer: OK. Resolving the conversation.

> \frac{d}{dt}f(x(t),y(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}+\frac{\partial f}{\partial o}\frac{do}{dt}\\

Reviewer: `o` is missing:
```suggestion
\frac{d}{dt}f(x(t),y(t),o(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}+\frac{\partial f}{\partial o}\frac{do}{dt}\\
```

&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\cdot 0+\frac{\partial f}{\partial o}\cdot 0\\

> &=&\frac{\partial f}{\partial x}\frac{dx}{dt}

Reviewer: Should we include `f_der` into the formula? 
```suggestion
&=&\frac{\partial f}{\partial x}\frac{dx}{dt}\\
&=&f\_der
```
Or it can be on the same line?

Author: I would prefer not to include `f_der` at this time, as the underscore looks misplaced in the pdf-file. 
The next option would be to rename f_der in this case as well (e.g. to h) - which should work although a bit odd.

Author: I renamed to fDer, and included it.

\end{lstlisting}
This is useful if \lstinline!g! represents the major computational
effort of \lstinline!fg!.

Therefore \lstinline!f_der! indirectly includes the derivative with respect to \lstinline!y! as follows:
\begin{eqnarray*}
\frac{d}{dt}fg(x(t))&=&\frac{d}{dt}f(x(t),g(x(t)))\\
&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{\partial g}{\partial x}\frac{dx}{dt}\\

> &=&(\frac{\partial f}{\partial x}+\frac{\partial f}{\partial y}\frac{\partial g}{\partial x})\frac{dx}{dt}\\

Reviewer: Might be better to use adjustable parentheses:
```suggestion
&=&\left(\frac{\partial f}{\partial x}+\frac{\partial f}{\partial y}\frac{\partial g}{\partial x}\right)\frac{dx}{dt}\\
```

> &=&\textrm{f\_der}(x,y)\frac{dx}{dt}

Author: Since `f_der` is not merely the derivative of `f` I renamed it to `h`; and that also corrected the font-issue.
